walkeeper/src/wal_service.rs

```diff
@@ -984,24 +987,29 @@ impl Connection {
             } else {
                 /* normal mode */
                 let timeline = self.timeline();
+                let mut shared_state = timeline.mutex.lock().unwrap();
                 loop {
-                    // Rust doesn't allow to grab async result from mutex scope
-                    {
-                        let shared_state = timeline.mutex.lock().unwrap();
-                        commit_lsn = shared_state.commit_lsn;
-                        if start_pos < commit_lsn {
-                            end_pos = commit_lsn;
-                            break;
-                        }
+                    commit_lsn = shared_state.commit_lsn;
+                    if start_pos < commit_lsn {
+                        end_pos = commit_lsn;
+                        break;
                     }
-                    timeline.cond.notified().await;
+                    shared_state = timeline.cond.wait(shared_state).unwrap();
                 }
             }
             if end_pos == END_REPLICATION_MARKER {
                 break;
             }
             // Try to fetch replica's feedback
-            match self.stream.try_read_buf(&mut self.inbuf) {
+
+            // Temporarily set this stream into nonblocking mode.
+            // FIXME: This seems like a dirty hack.
+            // Should this task be done on a background thread?
+            self.stream.set_nonblocking(true).unwrap();
+            let read_result = self.stream.read(&mut self.inbuf);
+            self.stream.set_nonblocking(false).unwrap();
+
```

Comment:
I really hate this part too.

I don't have a better plan yet. I looked at using `nix` to call `select` or `poll` on the socket fd to test if it was readable, but it would have been slightly more ugly code, and we would still have 2 system calls (select, recv) rather than 3 (fcntl, recv, fcntl). I could go either way.

I'm not sure what our strategy should be here: should we spawn a background thread? If we do choose to keep the nonblocking socket I/O, we might want to hide the ugly details behind some helper trait or wrapper type.

Reply:
Hmm, the logic is a bit unclear to me anyway. We're blocking earlier in the loop, waiting on the Notify/CondVar. But shouldn't we process incoming HotStandbyFeedback messages as soon as they come in, even if we have nothing to send yet?

To be honest, tokio-async would feel natural here, but a separate thread is fine, too. In either case, I think the logic needs some love and comments.

Reply:
I haven't spent any time trying to actually understand why this code does what it does. This is trying to copy the logic that was there before.

I agree there are issues with trying to juggle send/receive in a single thread; I'm not sure yet what the right solution is. I don't see a clear way to use async here that would prevent accidental calling of blocking functions. It might become clearer if I spend more time understanding the protocol logic here...